File: app/src/main/java/awais/instagrabber/repositories/UserRepository.java
package awais.instagrabber.repositories;

import awais.instagrabber.repositories.responses.UserSearchResponse;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface UserRepository {

    @GET("/api/v1/users/{uid}/info/")
    Call<String> getUserInfo(@Path("uid") final String uid);

    @GET("/api/v1/users/search/")
    Call<UserSearchResponse> search(@Query("timezone_offset") float timezoneOffset,
                                    @Query("q") String query);
}

File: app/src/main/java/awais/instagrabber/webservices/UserRepository.kt
package awais.instagrabber.webservices

import awais.instagrabber.repositories.UserService
import awais.instagrabber.repositories.responses.FriendshipStatus
import awais.instagrabber.repositories.responses.User
import awais.instagrabber.repositories.responses.UserSearchResponse
import awais.instagrabber.webservices.RetrofitFactory.retrofit
import java.util.*

open class UserRepository(private val service: UserService) {

    suspend fun getUserInfo(uid: Long): User {
        val response = service.getUserInfo(uid)
        return response.user
    }

    open suspend fun getUsernameInfo(username: String): User {
        val response = service.getUsernameInfo(username)
        return response.user
    }

    open suspend fun getUserFriendship(uid: Long): FriendshipStatus = service.getUserFriendship(uid)

    suspend fun search(query: String): UserSearchResponse {
        val timezoneOffset = TimeZone.getDefault().rawOffset.toFloat() / 1000
        return service.search(timezoneOffset, query)
    }

    companion object {
        @Volatile
        private var INSTANCE: UserRepository? = null

        fun getInstance(): UserRepository {
            return INSTANCE ?: synchronized(this) {
                val service: UserService = retrofit.create(UserService::class.java)
                UserRepository(service).also { INSTANCE = it }
            }
        }
    }
}